pkg/common/cns-lib/volume/manager.go

```diff
@@ -1014,3 +1018,73 @@ func (m *defaultManager) RetrieveVStorageObject(ctx context.Context, volumeID st
 	log.Debugf("vStorageObject for volumeID: %q is %+v", volumeID, vStorageObject)
 	return vStorageObject, nil
 }
+
+// QueryVolumeAsync returns volumes matching the given filter by using CnsQueryAsync API. QueryVolumeAsync takes querySelection spec which helps to specify which fields
+// for the query entities to be returned. All volume fields would be returned as part of the CnsQueryResult if the querySelection parameters are not specified
+func (m *defaultManager) QueryVolumeAsync(ctx context.Context, queryFilter cnstypes.CnsQueryFilter, querySelection cnstypes.CnsQuerySelection) (*cnstypes.CnsQueryResult, error) {
+	internalQueryVolumeAsync := func() (*cnstypes.CnsQueryResult, error) {
+		log := logger.GetLogger(ctx)
+		isvSphere70U3orAbove, err := cnsvsphere.IsvSphereVersion70U3orAbove(ctx, m.virtualCenter.Client.ServiceContent.About)
+		if err != nil {
+			msg := fmt.Sprintf("Error while checking the vSphere Version %q to invoke QueryVolumeAsync, Error= %+v", m.virtualCenter.Client.ServiceContent.About.ApiVersion, err)
+			log.Errorf(msg)
+			return nil, errors.New(msg)
+		}
+		if !isvSphere70U3orAbove {
```

Comment:
If this fails, we do want to default to `QueryVolume` API? How are clients going to handle this?

Reply:
Yes, the callers will check for vsphere version and fallback to `QueryVolume` (Sync) API.

Reply:
Actually this is an extra check if callers do not perform a vSphere version check. The caller should fallback to `QueryVolume` (Sync) API for vsphere <=7.0u3